src/bucket/BucketIndexImpl.cpp

```diff
@@ -0,0 +1,531 @@
+                params.projected_element_count,
+                params.false_positive_probability,
+                params.optimal_parameters.number_of_hashes,
+                params.optimal_parameters.table_size);
+        }
+        else
+        {
+            estimatedIndexEntries = estimatedNumElems;
+        }
+
+        mData.keysToOffset.reserve(estimatedIndexEntries);
+
+        XDRInputFileStream in;
+        in.open(filename.string());
+        std::streamoff pos = 0;
+        std::streamoff pageUpperBound = 0;
+        BucketEntry be;
+        size_t iter = 0;
+        size_t count = 0;
+        while (in && in.readOne(be))
+        {
+            // peridocially check if bucket manager is exiting to stop indexing
+            // gracefully
+            if (++iter >= 1000)
+            {
+                iter = 0;
+                if (bm.isShutdown())
+                {
+                    throw std::runtime_error("Incomplete bucket index due to "
+                                             "BucketManager shutdown");
+                }
+            }
+
+            if (be.type() != METAENTRY)
+            {
+                ++count;
+                LedgerKey key = getBucketLedgerKey(be);
+                if constexpr (std::is_same<IndexT, RangeIndex>::value)
+                {
+                    if (pos >= pageUpperBound)
+                    {
+                        pageUpperBound =
+                            roundDown(pos, mData.pageSize) + mData.pageSize;
+                        mData.keysToOffset.emplace_back(RangeEntry(key, key),
+                                                        pos);
+                    }
+                    else
+                    {
+                        auto& rangeEntry = mData.keysToOffset.back().first;
+                        releaseAssert(rangeEntry.upperBound < key);
+                        rangeEntry.upperBound = key;
+                    }
+
+                    mData.filter->insert(std::hash<stellar::LedgerKey>()(key));
+                }
+                else
+                {
+                    mData.keysToOffset.emplace_back(key, pos);
+                }
+            }
+
+            pos = in.pos();
+        }
+
+        CLOG_DEBUG(Bucket, "Indexed {} positions in {}",
+                   mData.keysToOffset.size(), filename.filename());
+        if (std::is_same<IndexT, RangeIndex>::value &&
+            estimatedNumElems < count)
+        {
+            CLOG_WARNING(Bucket,
+                         "Underestimated bloom filter size. Estimated entry "
+                         "count: {}, Actual: {}",
+                         estimatedNumElems, count);
+        }
+        ZoneValue(static_cast<int64_t>(count));
+    }
+
+    if (bm.getConfig().EXPERIMENTAL_BUCKETLIST_DB_PERSIST_INDEX)
+    {
+        auto timer = LogSlowExecution("Saving index");
+        std::filesystem::path indexFilename =
+            bm.bucketFilename(hash, /*isIndex=*/true);
```

Comment:
posting before I forget: we should probably do the whole tmp file + durable rename dance here instead of writing straight to the buckets folder, so that we don't leave invalid files due to a crash

Reply:
(Marking resolved as Marta is on vacation and I think this has been addressed)